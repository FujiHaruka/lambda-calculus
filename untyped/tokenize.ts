import { TokenizerContext } from "./tokenizer/TokenizerContext.ts";
import { Token } from "./tokenizer/types.ts";
import { scanSymbol } from "./tokenizer/scanSymbol.ts";
import { scanVar } from "./tokenizer/scanVar.ts";
import { UnexpectedCharacterError } from "./tokenizer/errors.ts";

function scan(ctx: TokenizerContext): Token | null {
  return scanSymbol(ctx) || scanVar(ctx);
}

export type TokenizeOptions = {
  eofToken?: boolean
}

/**
 * Tokenize the given code.
 */
export function tokenize(code: string, options: TokenizeOptions = {}): Token[] {
  const ctx = new TokenizerContext({ code, position: 0 });

  ctx.skipSpaces();
  while (!ctx.isEOF) {
    const maybeToken = scan(ctx);
    if (maybeToken) {
      ctx.pushToken(maybeToken);
    } else {
      throw new UnexpectedCharacterError(ctx);
    }
    ctx.skipSpaces();
  }

  if (options.eofToken) {
    return ctx.tokens.concat({
      type: "eof",
      start: ctx.position,
      end: ctx.position,
    })
  } else {
    return ctx.tokens;
  }
}
