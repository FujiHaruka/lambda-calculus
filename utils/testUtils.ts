export {
  assertEquals,
  assertStrictEquals,
  assertStringIncludes,
  assertThrows,
} from "https://deno.land/std@0.192.0/testing/asserts.ts";
export { describe, it } from "https://deno.land/std@0.192.0/testing/bdd.ts";
export { assertSnapshot } from "https://deno.land/std@0.192.0/testing/snapshot.ts";
export { walkSync } from "https://deno.land/std@0.192.0/fs/mod.ts";
export * as path from "https://deno.land/std@0.192.0/path/mod.ts";
