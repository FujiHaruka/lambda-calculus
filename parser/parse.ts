import { Token } from "../tokenizer/types.ts";
import { Stack } from "./Stack.ts";
import { ParenthesisNotClosedError, UnexpectedTokenError } from "./errors.ts";
import {
  AbstractionNode,
  ApplicationNode,
  Node,
  VariableNode,
} from "./types.ts";
import { tokenize } from "../tokenizer/tokenize.ts";
import { PartialNode } from "./PartialNode.ts";

export function parse(code: string): Node {
  const tokens = tokenize(code, { eofToken: true });
  const value = (token: Token): string => code.slice(token.start, token.end);
  const stack = new Stack<PartialNode>();
  stack.push(
    new PartialNode({
      type: "root",
    }, {
      leftParen: false,
    }),
  );

  for (const token of tokens) {
    const node = stack.top();

    if (
      // e.g.
      // "a b" + "c"
      // "(a b" + "c"
      token.type === "var" &&
      node &&
      node.type !== "abstraction" &&
      node.isNode() &&
      !node.rightParen
    ) {
      stack.pop();
      const nextNode = new PartialNode({
        type: "application",
        left: node.toNode(),
        right: {
          type: "var",
          identifier: value(token),
        },
      }, {
        leftParen: node.leftParen,
      });
      stack.push(nextNode);
    } else if (
      // e.g.
      // "a -> b" + "c"
      // "(a -> b" + "c"
      token.type === "var" &&
      node &&
      node.type === "abstraction" &&
      node.isNode() &&
      !node.rightParen
    ) {
      stack.pop();
      const abstractionNode = node.toNode() as AbstractionNode;
      stack.push(
        new PartialNode({
          type: "abstraction",
          bound: abstractionNode.bound,
        }, {
          leftParen: node.leftParen,
        }),
      );
      stack.push(
        new PartialNode({
          type: "application",
          left: abstractionNode.body,
          right: {
            type: "var",
            identifier: value(token),
          },
        }, {
          leftParen: false,
        }),
      );
    } else if (
      // e.g.
      // "(" + "a"
      // "(a ->" + "b"
      // "((a b)" + "c"
      token.type === "var" &&
      node &&
      node.type !== "root" &&
      !node.hasChild()
    ) {
      node.setChild({
        type: "var",
        identifier: value(token),
      });
    } else if (
      // e.g.
      // "" + "a"
      token.type === "var" &&
      node &&
      node.type === "root"
    ) {
      const nextNode = new PartialNode({
        type: "any",
        child: {
          type: "var",
          identifier: value(token),
        },
      }, {
        leftParen: false,
      });
      stack.push(nextNode);
    } else if (
      // e.g.
      // "(a" + "b"
      token.type === "var" &&
      node &&
      node.type === "any" &&
      node.hasChild()
    ) {
      stack.pop();
      stack.push(
        new PartialNode({
          type: "application",
          left: node.child!,
          right: {
            type: "var",
            identifier: value(token),
          },
        }, {
          leftParen: node.leftParen,
        }),
      );
    } else if (
      // e.g.
      // "(a b)" + "c"
      // "(a -> b)" + "c"
      token.type === "var" &&
      node &&
      node.isNode() &&
      node.rightParen
    ) {
      stack.pop();
      const nextNode = new PartialNode({
        type: "application",
        left: node.toNode(),
        right: {
          type: "var",
          identifier: value(token),
        },
      }, {
        leftParen: false,
      });
      stack.push(nextNode);
    } else if (
      // e.g.
      // "a b" + "("
      token.type === "left_paren" &&
      node &&
      node.type === "application" &&
      node.isNode() &&
      !node.rightParen
    ) {
      stack.pop();
      const nextNode = new PartialNode({
        type: "application",
        left: node.toNode(),
      }, {
        leftParen: node.leftParen,
      });
      stack.push(nextNode);

      const nextAnyNode = new PartialNode({
        type: "any",
      }, {
        leftParen: true,
      });
      stack.push(nextAnyNode);
    } else if (
      // e.g.
      // "(a b)" + "("
      // "(a -> b)" + "("
      token.type === "left_paren" &&
      node &&
      node.isNode() &&
      node.rightParen
    ) {
      stack.pop();
      const nextNode = new PartialNode({
        type: "application",
        left: node.toNode(),
      }, {
        leftParen: false,
      });
      stack.push(nextNode);

      const nextAnyNode = new PartialNode({
        type: "any",
      }, {
        leftParen: true,
      });
      stack.push(nextAnyNode);
    } else if (
      // e.g.
      // "("
      token.type === "left_paren"
    ) {
      const nextNode = new PartialNode({
        type: "any",
      }, {
        leftParen: true,
      });
      stack.push(nextNode);
    } else if (
      // e.g.
      // "(a -> b" + ")"
      // "((a b) c" + ")"
      // "(a (b c" + ")"
      token.type === "right_paren" &&
      node &&
      node.isNode() &&
      !node.isComplete()
    ) {
      stack.pop();
      const top = stack.pop();
      switch (top.type) {
        case "abstraction": {
          if (top.hasChild()) {
            // e.g.
            // "x -> y (z w" + ")"
            const abstractionNode = top.toNode() as AbstractionNode;
            const applicationNode: ApplicationNode = {
              type: "application",
              left: top.child!,
              right: node.toNode(),
            };
            const nextNode = new PartialNode({
              type: "abstraction",
              bound: abstractionNode.bound,
              body: applicationNode,
            }, {
              leftParen: top.leftParen,
            });
            stack.push(nextNode);
          } else {
            // e.g.
            // "(x -> (y -> z" + ")"
            top.setChild(node.toNode());
            stack.push(top);
          }
          break;
        }
        case "application": {
          // e.g.
          // "(x (y z" + ")"
          top.setChild(node.toNode());
          stack.push(top);
          break;
        }
        case "any": {
          const children = top.hasChild()
            // e.g.
            // "(a (b c" + ")"
            ? {
              left: top.child!,
              right: node.toNode(),
            }
            // e.g.
            // "((x y" + ")"
            : {
              left: node.toNode(),
            };
          stack.push(
            new PartialNode({
              type: "application",
              ...children,
            }, {
              leftParen: top.leftParen,
            }),
          );
          break;
        }
        case "root": {
          node.rightParen = true;
          stack.push(node);
          break;
        }
      }
    } else if (
      // e.g.
      // "(a -> b -> c" + ")"
      // "(a -> b -> c -> d" + ")"
      // "(x -> y z" + ")"
      token.type === "right_paren" &&
      node &&
      node.isComplete() &&
      stack.size >= 2
    ) {
      let childNode = node.toNode();
      stack.pop();

      const top = stack.top();
      while (
        top &&
        top.type === "abstraction" &&
        !top.hasChild() &&
        !top.leftParen
      ) {
        top.setChild(childNode);
        childNode = top.toNode();
        stack.pop();
      }

      const nextNode = stack.pop();
      if (
        nextNode &&
        nextNode.type === "abstraction" &&
        !nextNode.hasChild() &&
        nextNode.leftParen
      ) {
        nextNode.setChild(childNode);
        nextNode.rightParen = true;
      } else {
        throw new UnexpectedTokenError({ token, code });
      }

      const anyNode = stack.top();
      if (
        anyNode &&
        anyNode.type === "any" &&
        !anyNode.hasChild()
      ) {
        stack.pop();
        stack.push(
          new PartialNode({
            type: "application",
            left: nextNode.toNode(),
          }, {
            leftParen: anyNode.leftParen,
          }),
        );
      } else {
        stack.push(nextNode);
      }
    } else if (
      // e.g.
      // "(a" + "->"
      token.type === "arrow" &&
      node && node.type === "any" && node.child && node.child.type === "var"
    ) {
      stack.pop();
      stack.push(
        new PartialNode({
          type: "abstraction",
          bound: node.child.identifier,
        }, {
          leftParen: node.leftParen,
        }),
      );
    } else if (
      // e.g.
      // "a -> b" + "->"
      // "(a -> b" + "->"
      token.type === "arrow" &&
      node &&
      node.type === "abstraction" &&
      node.isNode() &&
      node.child?.type === "var" &&
      !node.rightParen
    ) {
      const abstractionNode = node.toNode() as AbstractionNode;
      const body = abstractionNode.body as VariableNode;
      stack.pop();
      stack.push(
        new PartialNode({
          type: "abstraction",
          bound: abstractionNode.bound,
        }, {
          leftParen: node.leftParen,
        }),
      );
      stack.push(
        new PartialNode({
          type: "abstraction",
          bound: body.identifier,
        }, {
          leftParen: false,
        }),
      );
    } else if (
      // e.g.
      // "a -> b" + "EOF"
      token.type === "eof" &&
      node &&
      !node.leftParen &&
      node.isNode()
    ) {
      let resultNode = node.toNode();
      stack.pop();
      // Reduce until root node
      while (!stack.empty() && stack.top()?.type !== "root") {
        const node = stack.pop();
        if (node.isComplete()) {
          throw new Error("Unexpected error");
        }
        node.setChild(resultNode);
        resultNode = node.toNode();
      }
      return resultNode;
    } else if (
      // e.g.
      // "(a -> b)" + "EOF"
      token.type === "eof" &&
      node &&
      node.isComplete()
    ) {
      return node.toNode();
    } else if (
      // e.g.
      // "a" + "EOF"
      token.type === "eof" &&
      node &&
      node.type === "any" &&
      !node.leftParen &&
      node.child
    ) {
      return node.child;
    } else if (
      // e.g.
      // "(a -> b" + "EOF"
      token.type === "eof" &&
      node &&
      !node.isComplete()
    ) {
      throw new ParenthesisNotClosedError();
    } else {
      throw new UnexpectedTokenError({ token, code });
    }
  }

  throw new Error("Unexpected error");
}
