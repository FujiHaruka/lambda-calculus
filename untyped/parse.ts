import { assertNever } from "./utils.ts";
import { Token } from "./tokenizer/types.ts";
import { Stack } from "./parser/Stack.ts";
import {
  ParenthesisNotClosedError,
  UnexpectedTokenError,
} from "./parser/errors.ts";
import { Node, VariableNode } from "./parser/types.ts";
import { tokenize } from "./tokenize.ts";
import { PartialNode } from "./parser/PartialNode.ts";

export function parse(code: string): Node {
  const tokens = tokenize(code);
  const value = (token: Token): string => code.slice(token.start, token.end);
  const stack = new Stack<PartialNode>();
  let rootNode: Node | null = null;

  for (const token of tokens) {
    const node = stack.top();

    switch (token.type) {
      case "var": {
        if (!node) {
          // This is a special case where the first token is a variable.
          // TODO: handle edge cases like there are more tokens after this one.
          return {
            type: "var",
            identifier: value(token),
          };
        }

        const varNode: VariableNode = {
          type: "var",
          identifier: value(token),
        };
        switch (node.type) {
          case "abstraction":
          case "application": {
            if (node.hasChild()) {
              throw new UnexpectedTokenError(token, value(token));
            }
            node.setChild(varNode);
            break;
          }
          case "any": {
            if (node.hasChild()) {
              node.evolve({
                type: "application",
                left: node.child!,
                right: varNode,
              });
            } else {
              node.setChild(varNode);
            }
            break;
          }
          default: {
            assertNever(node.type);
          }
        }
        break;
      }
      case "left_paren": {
        const nextNode = new PartialNode({
          type: "any",
        });
        stack.push(nextNode);
        break;
      }
      case "right_paren": {
        // The top node is expected to be a complete node, not a partial one.
        // e.g. In the second right parenthesis in "((a b) c)", the top node is "a b" and it's a complete node of type "application".
        if (!node || node.isImcompleteNode()) {
          throw new UnexpectedTokenError(token, value(token));
        }

        stack.pop();

        const top = stack.top();
        if (!top) {
          // Empty stack means we are done parsing
          rootNode = node.toNode();
          break;
        }

        switch (top.type) {
          case "abstraction":
          case "application": {
            if (top.hasChild()) {
              throw new UnexpectedTokenError(token, value(token));
            }

            top.setChild(node.toNode());
            break;
          }
          case "any": {
            if (top.hasChild()) {
              top.evolve({
                type: "application",
                left: top.child!,
                right: node.toNode(),
              });
            } else {
              top.evolve({
                type: "application",
                left: node.toNode(),
              });
            }
          }
        }
        break;
      }
      case "arrow": {
        if (
          node && node.type === "any" && node.child && node.child.type === "var"
        ) {
          node.evolve({
            type: "abstraction",
            bound: node.child.identifier,
          });
        } else {
          throw new UnexpectedTokenError(token, value(token));
        }
        break;
      }
      default: {
        assertNever(token.type);
      }
    }
  }

  if (!stack.empty()) {
    throw new ParenthesisNotClosedError();
  }
  if (!rootNode) {
    // This should never happen
    throw new Error("Root node is null");
  }

  return rootNode;
}
