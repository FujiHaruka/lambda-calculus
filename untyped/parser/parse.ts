import { assertNever } from "../utils.ts";
import { Token } from "../tokenizer/types.ts";
import { Stack } from "./Stack.ts";
import { UnexpectedTokenError } from "./errors.ts";
import { Node, PartialNode, VariableNode } from "./types.ts";

export function parse(code: string, tokens: Token[]): Node {
  const value = (token: Token): string => code.slice(token.start, token.end);
  const stack = new Stack<PartialNode>();

  for (const token of tokens) {
    const node = stack.top();

    if (node) {
      switch (token.type) {
        case "var": {
          const varNode: VariableNode = {
            type: "var",
            identifier: value(token),
          };
          switch (node.type) {
            case "var": {
              throw new UnexpectedTokenError(token, value(token));
            }
            case "abstraction": {
              if (node.body) {
                throw new UnexpectedTokenError(token, value(token));
              }
              node.body = varNode;
              break;
            }
            case "application": {
              if (node.right) {
                throw new UnexpectedTokenError(token, value(token));
              }
              node.right = varNode;
              break;
            }
            case "any": {
              if (node.child) {
                throw new UnexpectedTokenError(token, value(token));
              }
              node.child = varNode;
              break;
            }
            default: {
              assertNever(node);
            }
          }
          break;
        }
        case "left_paren": {
          stack.push({
            type: "any",
          });
          break;
        }
        case "right_paren": {
          break;
        }
        case "arrow": {
          if (node.type === "any" && node.child && node.child.type === "var") {
            stack.replaceTop({
              type: "abstraction",
              bound: node.child.identifier,
            });
          } else {
            throw new UnexpectedTokenError(token, value(token));
          }
          break;
        }
        default: {
          assertNever(token.type);
        }
      }
    } else {
      // First token
      switch (token.type) {
        case "var": {
          stack.push({
            type: "var",
            identifier: value(token),
          });
          break;
        }
        case "left_paren": {
          stack.push({
            type: "any",
          });
          break;
        }
        case "right_paren": {
          throw new UnexpectedTokenError(token, value(token));
        }
        case "arrow": {
          throw new UnexpectedTokenError(token, value(token));
        }
        default: {
          assertNever(token.type);
        }
      }
    }
  }

  const node = stack.pop();
  if (!stack.empty()) {
    throw new Error("Stack is not empty");
  }
  if (node.type === "any") {
    throw new Error("Node is still any");
  }

  // TODO: Check if node is not partial
  return node as Node;
}
