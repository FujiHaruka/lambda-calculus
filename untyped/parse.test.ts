import { parse } from "./parse.ts";
import { assertSnapshot, it } from "./testUtils.ts";
import { tokenize } from "./tokenize.ts";

[
  "x",
  "(x -> y)",
  "(x -> (y -> z))",
  "(x y)",
  "((x y) z)",
  "(x (y z))",
  "(((x y) (a b)) (c -> d))",
  "((x -> (y -> x)) (y z))",
  "((x -> (y -> x)) y)",
  "(x ((y -> x) (a -> b)))",
  "((((x -> (y -> (z -> ((x z) (y z))))) a) (x -> (y -> x))) c)",
].forEach((code) => {
  it(`parse "${code}"`, async (t) => {
    const tokens = tokenize(code);
    await assertSnapshot(t, parse(code, tokens));
  });
});
