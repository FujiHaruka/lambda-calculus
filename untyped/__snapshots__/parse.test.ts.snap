export const snapshot = {};

snapshot[`parse "x" 1`] = `
{
  identifier: "x",
  type: "var",
}
`;

snapshot[`parse "(x -> y)" 1`] = `
{
  body: {
    identifier: "y",
    type: "var",
  },
  bound: "x",
  type: "abstraction",
}
`;

snapshot[`parse "(x -> (y -> z))" 1`] = `
{
  body: {
    body: {
      identifier: "z",
      type: "var",
    },
    bound: "y",
    type: "abstraction",
  },
  bound: "x",
  type: "abstraction",
}
`;

snapshot[`parse "(x y)" 1`] = `
{
  left: {
    identifier: "x",
    type: "var",
  },
  right: {
    identifier: "y",
    type: "var",
  },
  type: "application",
}
`;

snapshot[`parse "((x y) z)" 1`] = `
{
  left: {
    left: {
      identifier: "x",
      type: "var",
    },
    right: {
      identifier: "y",
      type: "var",
    },
    type: "application",
  },
  right: {
    identifier: "z",
    type: "var",
  },
  type: "application",
}
`;

snapshot[`parse "(x (y z))" 1`] = `
{
  left: {
    identifier: "x",
    type: "var",
  },
  right: {
    left: {
      identifier: "y",
      type: "var",
    },
    right: {
      identifier: "z",
      type: "var",
    },
    type: "application",
  },
  type: "application",
}
`;

snapshot[`parse "(((x y) (a b)) (c -> d))" 1`] = `
{
  left: {
    left: {
      left: {
        identifier: "x",
        type: "var",
      },
      right: {
        identifier: "y",
        type: "var",
      },
      type: "application",
    },
    right: {
      left: {
        identifier: "a",
        type: "var",
      },
      right: {
        identifier: "b",
        type: "var",
      },
      type: "application",
    },
    type: "application",
  },
  right: {
    body: {
      identifier: "d",
      type: "var",
    },
    bound: "c",
    type: "abstraction",
  },
  type: "application",
}
`;

snapshot[`parse "((x -> (y -> x)) (y z))" 1`] = `
{
  left: {
    body: {
      body: {
        identifier: "x",
        type: "var",
      },
      bound: "y",
      type: "abstraction",
    },
    bound: "x",
    type: "abstraction",
  },
  right: {
    left: {
      identifier: "y",
      type: "var",
    },
    right: {
      identifier: "z",
      type: "var",
    },
    type: "application",
  },
  type: "application",
}
`;

snapshot[`parse "((x -> (y -> x)) y)" 1`] = `
{
  left: {
    body: {
      body: {
        identifier: "x",
        type: "var",
      },
      bound: "y",
      type: "abstraction",
    },
    bound: "x",
    type: "abstraction",
  },
  right: {
    identifier: "y",
    type: "var",
  },
  type: "application",
}
`;

snapshot[`parse "(x ((y -> x) (a -> b)))" 1`] = `
{
  left: {
    identifier: "x",
    type: "var",
  },
  right: {
    left: {
      body: {
        identifier: "x",
        type: "var",
      },
      bound: "y",
      type: "abstraction",
    },
    right: {
      body: {
        identifier: "b",
        type: "var",
      },
      bound: "a",
      type: "abstraction",
    },
    type: "application",
  },
  type: "application",
}
`;

snapshot[`parse "((((x -> (y -> (z -> ((x z) (y z))))) a) (x -> (y -> x))) c)" 1`] = `
{
  left: {
    left: {
      left: {
        body: {
          body: {
            body: {
              left: {
                left: {
                  identifier: "x",
                  type: "var",
                },
                right: {
                  identifier: "z",
                  type: "var",
                },
                type: "application",
              },
              right: {
                left: {
                  identifier: "y",
                  type: "var",
                },
                right: {
                  identifier: "z",
                  type: "var",
                },
                type: "application",
              },
              type: "application",
            },
            bound: "z",
            type: "abstraction",
          },
          bound: "y",
          type: "abstraction",
        },
        bound: "x",
        type: "abstraction",
      },
      right: {
        identifier: "a",
        type: "var",
      },
      type: "application",
    },
    right: {
      body: {
        body: {
          identifier: "x",
          type: "var",
        },
        bound: "y",
        type: "abstraction",
      },
      bound: "x",
      type: "abstraction",
    },
    type: "application",
  },
  right: {
    identifier: "c",
    type: "var",
  },
  type: "application",
}
`;
